File: benchmark/oracles/trafilatura/selftest.py
#!/usr/bin/env python3
"""Committed dev-time self-test for the Trafilatura oracle adapter.

One command (see README). NOT a runtime dependency of run.py. Uses the
dev-time-only `jsonschema` validator installed into the same git-ignored
.venv (HLD section 6); not in the adapter's pinned runtime closure.

Implements the HLD section 7 five-step contract proof against
../contract.schema.json, plus a "found nothing != error" Bug-E2 guard. This
is the 'oracle of the oracles' for the Trafilatura side: it feeds known
inputs and asserts schema validity, the ok/error/exit tri-state, non-empty
text on a content page, the section 3.3 well-formed primitive, and same-machine
run-twice byte-identity.

Exit 0 iff every step passes; non-zero with a stderr reason otherwise.
"""

import json
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
RUN_PY = os.path.join(HERE, "run.py")
SCHEMA = os.path.join(HERE, "..", "contract.schema.json")
FIXTURES = os.path.join(HERE, "..", "fixtures")
ARTICLE = os.path.join(FIXTURES, "article.html")
EMPTY = os.path.join(FIXTURES, "empty.html")
UNICODE = os.path.join(FIXTURES, "unicode-article.html")
NONEXISTENT = os.path.join(FIXTURES, "this-path-does-not-exist.html")

# Exact non-Latin-1 code points the unicode-article fixture's MAIN content
# carries; cp1252 (the default Windows console/pipe codepage) cannot encode
# any of them. The HLD §3.3 contract mandates UTF-8 stdout independent of the
# host codepage; this is the permanent regression guard for that defect.
_UNICODE_REQUIRED = ("“", "”", "—", "Beyoncé", "µ", "日本語")


def _fail(step, msg):
    sys.stderr.write(f"[selftest.py] FAIL ({step}): {msg}\n")
    sys.exit(1)


def _invoke(path, *extra_args):
    """Invoke run.py exactly as the harness does: bare `python` + script +
    absolute path, plus any extra CLI args verbatim (e.g.
    ``"--base-url", "..."``). Returns (returncode, stdout_bytes,
    stderr_text)."""
    proc = subprocess.run(
        [sys.executable, RUN_PY, path, *extra_args],
        capture_output=True,
        check=False,
    )
    return proc.returncode, proc.stdout, proc.stderr.decode("utf-8", "replace")


def _load_schema():
    with open(SCHEMA, "r", encoding="utf-8") as fh:
        return json.load(fh)


def _validate(step, schema, obj):
    import jsonschema

    try:
        jsonschema.validate(instance=obj, schema=schema)
    except jsonschema.ValidationError as exc:  # type: ignore[attr-defined]
        _fail(step, f"schema validation failed: {exc.message}")


def _parse_stdout(step, stdout_bytes):
    try:
        return json.loads(stdout_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        _fail(step, f"stdout is not one valid UTF-8 JSON object: {exc}")


def _has_surrogate(text):
    return any(0xD800 <= ord(ch) <= 0xDFFF for ch in text)


def main():
    schema = _load_schema()

    # --- Step 1: article.html -> schema-valid, ok:true, exit 0, substantive
    #     text, no <head>/<script>/<style> leak, one clean object despite the
    #     fixture's deliberately malformed CSS rule. -----------------------
    rc, out, _err = _invoke(ARTICLE)
    if rc != 0:
        _fail("step1", f"expected exit 0 on article.html, got {rc}")
    obj = _parse_stdout("step1", out)
    _validate("step1", schema, obj)
    if obj.get("ok") is not True or obj.get("error") is not None:
        _fail("step1", f"expected ok:true/error:null, got {obj!r}")
    if not isinstance(obj.get("text"), str) or len(obj["text"].strip()) < 80:
        _fail("step1", "expected substantive `text` on the article fixture")
    leaked = "SCRIPT_SHOULD_NOT_APPEAR_IN_TEXT"
    if leaked in obj["text"]:
        _fail("step1", "<script> content leaked into extracted `text`")
    if "color: #123456" in obj["text"] or "@media screen" in obj["text"]:
        _fail("step1", "<style> content leaked into extracted `text`")
    if obj.get("contract_version") != 1:
        _fail("step1", f"contract_version must be 1, got {obj.get('contract_version')!r}")

    # --- Step 2: empty.html -> schema-valid, ok:true, text:"", exit 0
    #     (the Bug E2 'found nothing' guard). --------------------------------
    rc, out, _err = _invoke(EMPTY)
    if rc != 0:
        _fail("step2", f"expected exit 0 on empty.html, got {rc}")
    obj = _parse_stdout("step2", out)
    _validate("step2", schema, obj)
    if obj.get("ok") is not True or obj.get("error") is not None:
        _fail("step2", f"'found nothing' must be ok:true, got {obj!r}")
    if obj.get("text") != "":
        _fail("step2", f"expected text:'' on empty.html, got {obj.get('text')!r}")

    # --- Step 3: nonexistent path -> schema-valid failure envelope, fully
    #     field-determined per section 3.4: ok:false, error set, exit != 0
    #     (the Bug E2 'blew up — catchable' guard). --------------------------
    rc, out, _err = _invoke(NONEXISTENT)
    if rc == 0:
        _fail("step3", "expected non-zero exit on a nonexistent path")
    obj = _parse_stdout("step3", out)
    _validate("step3", schema, obj)
    if obj.get("ok") is not False or not obj.get("error"):
        _fail("step3", f"expected ok:false + error set, got {obj!r}")
    if obj.get("text") != "" or obj.get("word_count") is not None:
        _fail("step3", "failure envelope must be fully field-determined")
    if obj.get("contract_version") != 1 or obj.get("oracle") != "trafilatura":
        _fail("step3", "failure envelope must set contract_version & oracle")

    # --- Step 4: section 3.3 primitive guard. Construct a lone surrogate via a
    #     SOURCE-LEVEL escape (not a UTF-8 fixture file) and pass it through
    #     the same primitive run.py uses; assert the post-primitive text is
    #     schema-valid and contains NO code point in U+D800..U+DFFF. ---------
    lone = "before\ud800after"  # lone high surrogate via source escape.
    if not _has_surrogate(lone):
        _fail("step4", "test setup error: input had no lone surrogate")
    # Mirror run.py's pinned primitive exactly.
    cleaned = lone.encode("utf-8", "surrogatepass").decode("utf-8", "replace")
    if _has_surrogate(cleaned):
        _fail("step4", "post-primitive text still contains a lone surrogate")
    probe = {
        "contract_version": 1,
        "oracle": "trafilatura",
        "oracle_version": "2.0.0",
        "title": None,
        "text": cleaned,
        "html": None,
        "word_count": len(cleaned.split()),
        "canonical_url": None,
        "language": None,
        "ok": True,
        "error": None,
    }
    _validate("step4", schema, probe)
    # Round-trips through serde-compatible UTF-8 JSON without raising.
    json.loads(json.dumps(probe, ensure_ascii=False).encode("utf-8").decode("utf-8"))

    # --- Step 5: re-run step 1 and diff — byte-identical (same-machine
    #     determinism, section 3.5). -----------------------------------------
    _rc1, out1, _e1 = _invoke(ARTICLE)
    _rc2, out2, _e2 = _invoke(ARTICLE)
    if out1 != out2:
        _fail("step5", "two runs on article.html were NOT byte-identical")

    # --- Step 6: non-Latin-1 / Windows-codepage stdout regression guard.
    #     run.py writes its single JSON object to a PIPE (exactly how the Rust
    #     harness invokes it via Stdio::piped()); on Windows Python then sets
    #     sys.stdout.encoding to the ANSI codepage (cp1252), NOT UTF-8. The
    #     HLD §3.3 contract mandates UTF-8 stdout regardless of the host
    #     codepage. This fixture's MAIN content carries curly quotes, an
    #     em-dash, an accented name (Beyoncé), the micro sign (µ) and a CJK
    #     phrase (日本語) — every one cp1252-unencodable — so a codepage-
    #     dependent writer raises UnicodeEncodeError mid-emit and the whole
    #     contract envelope is lost (the corpus-wide oracle_error defect).
    #     Assert: exit 0, schema-valid, ok:true / error:null, the stdout
    #     bytes decode as UTF-8, and the decoded `text` contains EVERY
    #     required code point verbatim. `_invoke`'s capture_output=True gives
    #     run.py a pipe, faithfully reproducing the harness condition. -------
    rc, out, _err = _invoke(UNICODE)
    if rc != 0:
        _fail(
            "step6",
            f"expected exit 0 on unicode-article.html, got {rc} "
            f"(UTF-8 stdout contract violated; stderr below)\n{_err}",
        )
    # MUST decode as UTF-8 with strict errors (no replacement) — a cp1252 or
    # mojibake stream fails here, which is the precise defect being guarded.
    try:
        decoded = out.decode("utf-8")
    except UnicodeDecodeError as exc:
        _fail("step6", f"stdout is not valid UTF-8: {exc}")
    obj = _parse_stdout("step6", out)
    _validate("step6", schema, obj)
    if obj.get("ok") is not True or obj.get("error") is not None:
        _fail("step6", f"expected ok:true/error:null, got {obj!r}")
    if obj.get("contract_version") != 1:
        _fail(
            "step6",
            f"contract_version must be 1, got {obj.get('contract_version')!r}",
        )
    text = obj.get("text")
    if not isinstance(text, str) or len(text.strip()) < 80:
        _fail("step6", "expected substantive `text` on the unicode fixture")
    missing = [ch for ch in _UNICODE_REQUIRED if ch not in text]
    if missing:
        _fail(
            "step6",
            "extracted `text` is missing required non-Latin-1 code points "
            f"{missing!r} (U+%s) — UTF-8 stdout contract violated"
            % ", U+".join(f"{ord(c):04X}" for c in missing),
        )
    if "SCRIPT_SHOULD_NOT_APPEAR_IN_TEXT" in text:
        _fail("step6", "<script> content leaked into extracted `text`")
    # The decoded-bytes path and the json-parsed path must agree, proving the
    # code points survive the actual stdout byte stream (not just in-memory).
    if json.loads(decoded).get("text") != text:
        _fail("step6", "decoded-bytes `text` disagrees with parsed `text`")

    # --- Step 7: --base-url contract surface. --base-url is AUXILIARY
    #     (canonical/relative-link resolution only), NOT essential to
    #     main-content extraction. Trafilatura already tolerates a malformed
    #     url= and still extracts; this LOCKS that in as a permanent
    #     regression guard and keeps the two adapters symmetric on this
    #     surface (the sibling Readability adapter degrades a malformed
    #     --base-url to the no-base path rather than hard-failing). Assert all
    #     three contract cases: (a) VALID --base-url -> ok:true,
    #     byte-deterministic; (b) ABSENT --base-url -> ok:true (unchanged);
    #     (c) STRUCTURALLY-INVALID --base-url -> ok:true WITH substantive
    #     extracted text (graceful degrade, NOT a hard failure) AND
    #     byte-deterministic. run.py is NOT modified — it is already
    #     tolerant; (c) is expected to pass as-is. ---------------------------
    # (a) valid --base-url.
    rc, out_va, _err = _invoke(ARTICLE, "--base-url", "https://example.com/page")
    if rc != 0:
        _fail("step7", f"valid --base-url: expected exit 0, got {rc}")
    obj = _parse_stdout("step7", out_va)
    _validate("step7", schema, obj)
    if obj.get("ok") is not True or obj.get("error") is not None:
        _fail("step7", f"valid --base-url: expected ok:true, got {obj!r}")
    if not isinstance(obj.get("text"), str) or len(obj["text"].strip()) < 80:
        _fail("step7", "valid --base-url: expected substantive `text`")
    _rc, out_vb, _e = _invoke(ARTICLE, "--base-url", "https://example.com/page")
    if out_va != out_vb:
        _fail("step7", "valid --base-url: two runs were NOT byte-identical")

    # (b) absent --base-url (unchanged no-base behaviour).
    rc, out_ab, _err = _invoke(ARTICLE)
    if rc != 0:
        _fail("step7", f"absent --base-url: expected exit 0, got {rc}")
    obj = _parse_stdout("step7", out_ab)
    _validate("step7", schema, obj)
    if obj.get("ok") is not True or obj.get("error") is not None:
        _fail("step7", f"absent --base-url: expected ok:true, got {obj!r}")

    # (c) structurally-invalid --base-url: MUST still extract (ok:true, text
    #     present), MUST NOT escalate to the failure envelope, MUST be
    #     byte-deterministic. Trafilatura is already tolerant; lock it in.
    for bad in (
        "//x",
        "//proto-relative",
        "http://[invalid",
        "::::not a url::::",
        "http://a b c/x",
    ):
        rc, out_ia, err = _invoke(ARTICLE, "--base-url", bad)
        if rc != 0:
            _fail(
                "step7",
                f"invalid --base-url {bad!r}: expected exit 0 (graceful "
                f"degrade), got {rc}; stderr below\n{err}",
            )
        obj = _parse_stdout("step7", out_ia)
        _validate("step7", schema, obj)
        if obj.get("ok") is not True or obj.get("error") is not None:
            _fail(
                "step7",
                f"invalid --base-url {bad!r}: a malformed base is AUXILIARY "
                f"and must degrade gracefully, not hard-fail; got {obj!r}",
            )
        if not isinstance(obj.get("text"), str) or len(obj["text"].strip()) < 80:
            _fail(
                "step7",
                f"invalid --base-url {bad!r}: expected substantive `text` "
                "(extraction must still occur)",
            )
        _rc, out_ib, _e = _invoke(ARTICLE, "--base-url", bad)
        if out_ia != out_ib:
            _fail(
                "step7",
                f"invalid --base-url {bad!r}: two runs were NOT "
                "byte-identical",
            )

    sys.stderr.write(
        "[selftest.py] PASS — all 7 steps (schema-valid, tri-state, "
        "non-empty text, section 3.3 primitive, byte-identical re-run, "
        "non-Latin-1 UTF-8 stdout regression guard, --base-url "
        "valid/absent/invalid graceful-degrade).\n"
    )
    sys.exit(0)


if __name__ == "__main__":
    main()

File: benchmark/oracles/readability-js/selftest.mjs
// Committed dev-time self-test for the Readability-JS oracle adapter.
//
// One command (see README). NOT a runtime dependency of run.mjs. Uses the
// dev-time-only `ajv` validator from devDependencies (HLD section 6); not in
// the adapter's pinned runtime closure.
//
// Implements the HLD section 7 five-step contract proof against
// ../contract.schema.json, plus the "found nothing != error" Bug-E2 guard.
// The 'oracle of the oracles' for the Readability side: schema validity, the
// ok/error/exit tri-state, non-empty text on a content page, stdout-hygiene
// under the fixture's malformed CSS, the section 3.3 well-formed primitive,
// and same-machine run-twice byte-identity.
//
// Exit 0 iff every step passes; non-zero with a stderr reason otherwise.

import { spawnSync } from "node:child_process";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const HERE = dirname(fileURLToPath(import.meta.url));
const RUN_MJS = join(HERE, "run.mjs");
const SCHEMA = join(HERE, "..", "contract.schema.json");
const FIXTURES = join(HERE, "..", "fixtures");
const ARTICLE = join(FIXTURES, "article.html");
const EMPTY = join(FIXTURES, "empty.html");
const NONEXISTENT = join(FIXTURES, "this-path-does-not-exist.html");

function fail(step, msg) {
  process.stderr.write(`[selftest.mjs] FAIL (${step}): ${msg}\n`);
  process.exit(1);
}

// Invoke run.mjs exactly as the harness does: bare `node` + script + abs path,
// plus any extra CLI args verbatim (e.g. ["--base-url", "..."]).
function invoke(path, ...extraArgs) {
  const r = spawnSync(process.execPath, [RUN_MJS, path, ...extraArgs], {
    encoding: "buffer",
  });
  return {
    code: r.status,
    stdout: r.stdout,
    stderr: r.stderr ? r.stderr.toString("utf-8") : "",
  };
}

function loadValidator() {
  const Ajv = require("ajv");
  const schema = JSON.parse(readFileSync(SCHEMA, "utf-8"));
  const ajv = new Ajv({ allErrors: true, strict: false });
  return ajv.compile(schema);
}

function parseStdout(step, stdoutBuf) {
  let txt;
  try {
    txt = stdoutBuf.toString("utf-8");
  } catch (e) {
    fail(step, `stdout is not valid UTF-8: ${e}`);
  }
  try {
    return JSON.parse(txt);
  } catch (e) {
    fail(step, `stdout is not one valid JSON object: ${e} :: ${txt.slice(0, 200)}`);
  }
  return undefined;
}

function validate(step, validateFn, obj) {
  if (!validateFn(obj)) {
    fail(
      step,
      `schema validation failed: ${JSON.stringify(validateFn.errors)}`,
    );
  }
}

function hasSurrogate(s) {
  for (const ch of s) {
    const c = ch.codePointAt(0);
    if (c >= 0xd800 && c <= 0xdfff) return true;
  }
  return false;
}

const validateFn = loadValidator();

// --- Step 1: article.html -> schema-valid, ok:true, exit 0, substantive
//     text, no <head>/<script>/<style> leak, ONE clean object despite the
//     fixture's deliberately malformed CSS (exercises section 5 stderr-only
//     virtual console). ----------------------------------------------------
{
  const r = invoke(ARTICLE);
  if (r.code !== 0) fail("step1", `expected exit 0, got ${r.code}`);
  const obj = parseStdout("step1", r.stdout);
  validate("step1", validateFn, obj);
  if (obj.ok !== true || obj.error !== null) {
    fail("step1", `expected ok:true/error:null, got ${JSON.stringify(obj)}`);
  }
  if (typeof obj.text !== "string" || obj.text.trim().length < 80) {
    fail("step1", "expected substantive `text` on the article fixture");
  }
  if (obj.text.includes("SCRIPT_SHOULD_NOT_APPEAR_IN_TEXT")) {
    fail("step1", "<script> content leaked into extracted `text`");
  }
  if (obj.text.includes("color: #123456") || obj.text.includes("@media screen")) {
    fail("step1", "<style> content leaked into extracted `text`");
  }
  if (obj.contract_version !== 1) {
    fail("step1", `contract_version must be 1, got ${obj.contract_version}`);
  }
  // stdout must be exactly ONE object: parseStdout above already proved the
  // whole stdout is a single JSON value (no trailing parser noise).
}

// --- Step 2: empty.html -> schema-valid, ok:true, text:"", exit 0 (the Bug
//     E2 'found nothing' guard). example.com-style empties legitimately
//     return empty for a correct Readability adapter. ----------------------
{
  const r = invoke(EMPTY);
  if (r.code !== 0) fail("step2", `expected exit 0, got ${r.code}`);
  const obj = parseStdout("step2", r.stdout);
  validate("step2", validateFn, obj);
  if (obj.ok !== true || obj.error !== null) {
    fail("step2", `'found nothing' must be ok:true, got ${JSON.stringify(obj)}`);
  }
  if (obj.text !== "") {
    fail("step2", `expected text:'' on empty.html, got ${JSON.stringify(obj.text)}`);
  }
}

// --- Step 3: nonexistent path -> schema-valid failure envelope, fully
//     field-determined per section 3.4: ok:false, error set, exit != 0 (the
//     Bug E2 'blew up — catchable' guard). ---------------------------------
{
  const r = invoke(NONEXISTENT);
  if (r.code === 0) fail("step3", "expected non-zero exit on a nonexistent path");
  const obj = parseStdout("step3", r.stdout);
  validate("step3", validateFn, obj);
  if (obj.ok !== false || !obj.error) {
    fail("step3", `expected ok:false + error set, got ${JSON.stringify(obj)}`);
  }
  if (obj.text !== "" || obj.word_count !== null) {
    fail("step3", "failure envelope must be fully field-determined");
  }
  if (obj.contract_version !== 1 || obj.oracle !== "readability-js") {
    fail("step3", "failure envelope must set contract_version & oracle");
  }
}

// --- Step 4: section 3.3 primitive guard. Lone surrogate via a SOURCE-LEVEL
//     escape (not a UTF-8 fixture file); assert the post-primitive text is
//     schema-valid and contains NO code point in U+D800..U+DFFF. -----------
{
  const lone = "before\uD800after"; // lone high surrogate via source escape.
  if (!hasSurrogate(lone)) fail("step4", "test setup error: no lone surrogate");
  const cleaned = lone.toWellFormed(); // run.mjs's pinned primitive exactly.
  if (hasSurrogate(cleaned)) {
    fail("step4", "post-primitive text still contains a lone surrogate");
  }
  const probe = {
    contract_version: 1,
    oracle: "readability-js",
    oracle_version: "0.6.0",
    title: null,
    text: cleaned,
    html: null,
    word_count: cleaned.split(/\s+/u).length,
    canonical_url: null,
    language: null,
    ok: true,
    error: null,
  };
  validate("step4", validateFn, probe);
  JSON.parse(JSON.stringify(probe)); // round-trips without raising.
}

// --- Step 5: re-run step 1 and diff — byte-identical (same-machine
//     determinism, section 3.5). -------------------------------------------
{
  const a = invoke(ARTICLE);
  const b = invoke(ARTICLE);
  if (Buffer.compare(a.stdout, b.stdout) !== 0) {
    fail("step5", "two runs on article.html were NOT byte-identical");
  }
}

// --- Step 6: --base-url contract surface. --base-url is AUXILIARY
//     (canonical/relative-link resolution only), NOT essential to main-content
//     extraction; jsdom rejects an invalid `url` with `TypeError: Invalid
//     URL`. Without this guard a malformed --base-url would silently remove
//     THIS oracle from the differential for every malformed-URL page while the
//     Trafilatura oracle keeps participating (asymmetric instrument bias).
//     Assert all three contract cases: (a) VALID --base-url -> ok:true,
//     byte-deterministic; (b) ABSENT --base-url -> ok:true (the existing
//     no-base behaviour, unchanged); (c) STRUCTURALLY-INVALID --base-url ->
//     ok:true WITH substantive extracted text (graceful degrade, NOT a hard
//     failure) AND byte-deterministic. (b) and (c) must be byte-identical:
//     a malformed base must collapse to exactly the no-base path. -----------
{
  // (a) valid --base-url.
  const va = invoke(ARTICLE, "--base-url", "https://example.com/page");
  if (va.code !== 0) {
    fail("step6", `valid --base-url: expected exit 0, got ${va.code}`);
  }
  const vObj = parseStdout("step6", va.stdout);
  validate("step6", validateFn, vObj);
  if (vObj.ok !== true || vObj.error !== null) {
    fail(
      "step6",
      `valid --base-url: expected ok:true/error:null, got ${JSON.stringify(
        vObj,
      )}`,
    );
  }
  if (typeof vObj.text !== "string" || vObj.text.trim().length < 80) {
    fail("step6", "valid --base-url: expected substantive `text`");
  }
  const vb = invoke(ARTICLE, "--base-url", "https://example.com/page");
  if (Buffer.compare(va.stdout, vb.stdout) !== 0) {
    fail("step6", "valid --base-url: two runs were NOT byte-identical");
  }

  // (b) absent --base-url (unchanged no-base behaviour).
  const ab = invoke(ARTICLE);
  if (ab.code !== 0) {
    fail("step6", `absent --base-url: expected exit 0, got ${ab.code}`);
  }
  const abObj = parseStdout("step6", ab.stdout);
  validate("step6", validateFn, abObj);
  if (abObj.ok !== true || abObj.error !== null) {
    fail(
      "step6",
      `absent --base-url: expected ok:true/error:null, got ${JSON.stringify(
        abObj,
      )}`,
    );
  }

  // (c) structurally-invalid --base-url: MUST still extract (ok:true, text
  //     present), MUST NOT escalate to the ok:false/exit!=0 failure envelope,
  //     and MUST be byte-deterministic. Several malformed shapes.
  for (const bad of [
    "//x",
    "//proto-relative",
    "http://[invalid",
    "::::not a url::::",
    "http://a b c/x",
  ]) {
    const ia = invoke(ARTICLE, "--base-url", bad);
    if (ia.code !== 0) {
      fail(
        "step6",
        `invalid --base-url ${JSON.stringify(bad)}: expected exit 0 ` +
          `(graceful degrade), got ${ia.code} :: ${ia.stderr.slice(0, 200)}`,
      );
    }
    const iObj = parseStdout("step6", ia.stdout);
    validate("step6", validateFn, iObj);
    if (iObj.ok !== true || iObj.error !== null) {
      fail(
        "step6",
        `invalid --base-url ${JSON.stringify(bad)}: a malformed base is ` +
          `AUXILIARY and must degrade gracefully, not hard-fail; got ` +
          JSON.stringify(iObj),
      );
    }
    if (typeof iObj.text !== "string" || iObj.text.trim().length < 80) {
      fail(
        "step6",
        `invalid --base-url ${JSON.stringify(bad)}: expected substantive ` +
          "`text` (extraction must still occur)",
      );
    }
    const ib = invoke(ARTICLE, "--base-url", bad);
    if (Buffer.compare(ia.stdout, ib.stdout) !== 0) {
      fail(
        "step6",
        `invalid --base-url ${JSON.stringify(bad)}: two runs were NOT ` +
          "byte-identical",
      );
    }
    // A malformed base must collapse to EXACTLY the no-base path.
    if (Buffer.compare(ia.stdout, ab.stdout) !== 0) {
      fail(
        "step6",
        `invalid --base-url ${JSON.stringify(bad)}: output not ` +
          "byte-identical to the absent-base path (must be symmetric)",
      );
    }
  }
}

process.stderr.write(
  "[selftest.mjs] PASS — all 6 steps (schema-valid, tri-state, non-empty " +
    "text, section 3.3 primitive, byte-identical re-run, --base-url " +
    "valid/absent/invalid graceful-degrade).\n",
);
process.exit(0);
